Pipeline support

KamikaZeeFu opened the pull request:
Fixes #81 

This PR includes support for using this plugin in workflow jobs/pipelines.
The main needed change was refactoring of the actions. Using a `TransientActionFactory` makes sure that the action is only added once to a `Run`, no matter how many times the report step is run.
Actions have also been moved to a separate package, with old implementations marked as `@Deprecated`, so that old reports continue to work.

For a better user experience in pipeline scripts, the symbolic name `cucumber` was used for this plugin (analogous to the `junit` step from the JUnit plugin). 
Also most parameters have been marked as optional, since there are already sensible defaults set.
The only required parameter is `fileIncludePattern`.

Example usage:
```groovy
node('master') {
    try {
        //run some cucumber tests
        sh 'mvn clean install'
    } finally {
        //generate cucumber html report
        cucumber '**/cucumber-*.json'
    }
}
```

Current limitation is, that includes and excludes are only evaluated **after** the .json files are already copied to the cache directory. So if you call the step multiple times, only the includes and excludes of the very last invocation count.

A similar limitation is in effect for most other settings, as each invocation will generate a wholly new report.

KamikaZeeFu commented:
Squashed in the commit removing the checks for AbstractBuild, since the Token Macro plugin was updated.

damianszczepanik commented:
This PR looks like quite big refactor but not just simple support for pipeline. I'm happy to accept both but as separate changesets. Move all changes which are not related to pipeline into another PR, squash commits into one and send PR again.

KamikaZeeFu commented:
Sure thing, I'll update the PR accordingly.

damianszczepanik commented:
@KamikaZeeFu Hope you remember about your PR :)

kapoorlakshya commented:
Looking forward to this getting accepted and merged.

damianszczepanik commented:
First this PR has conflicts. Second it must be updated according to above comments

kapoorlakshya commented:
Right. Hoping that @KamikaZeeFu takes care of the issues.

KamikaZeeFu commented:
I updated the PR to only include pipeline relevant changes:

* Added 'cucumber' as symbolic name for pipeline usage
* The monolithic constructor was replaced by one only requiring the fileIncludePattern, other settings were moved to @DataboundSetter
    * This improves the generated snippets for pipeline, e.g. the generated snippet with all settings at default value turns 
    * from `cucumber  failedFeaturesNumber: 0, failedScenariosNumber: 0, failedStepsNumber: 0, fileExcludePattern: '', fileIncludePattern: '**/*.json', jsonReportDirectory: '', pendingStepsNumber: 0, skippedStepsNumber: 0, undefinedStepsNumber: 0` 
    * to `cucumber '**/*.json'`
* Logic for creating the cache directory had to be slightly adapted as pipelines can call a step any number of times (step would fail if cache directory already exists, which is the case when called twice)
* In the `perform` method, `run.addAction(caa)` was changed to `run.replaceAction(caa)` because multiple added actions will not work
    * This is the replacement hack for the action refactor removed from this PR.

damianszczepanik commented on README.md:
```diff
@@ -61,6 +61,31 @@ And you can drill down into tag specific reports:
 
 ![Tag report](https://github.com/damianszczepanik/cucumber-reporting/raw/master/.README/tag-report.png)
 
+### Pipeline usage
```
nice

damianszczepanik commented on src/main/java/net/masterthought/jenkins/CucumberReportPublisher.java:
```diff
@@ -1,37 +1,31 @@
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
```
I don't like this mess with imports

KamikaZeeFu replied:
That's the way my IDE formatted it, I'll try to revert it

damianszczepanik commented on src/main/java/net/masterthought/jenkins/CucumberReportPublisher.java:
```diff
@@ -83,26 +90,107 @@ public void setClassifications(List<Classification> classifications) {
+        this.buildStatus = buildStatus;
+    }
+
+    @DataBoundSetter
```
can you keep gettter and setter together?

KamikaZeeFu replied:
Sure

damianszczepanik commented on src/main/java/net/masterthought/jenkins/CucumberReportPublisher.java:
```diff
@@ -295,6 +379,7 @@ public String getDisplayName() {
     }
 
     @Extension
+    @Symbol("cucumber")
```
what is the reason for this?

KamikaZeeFu replied:
The @Symbol is what registers this plugin with the 'cucumber' pipeline command.

KamikaZeeFu commented:
Restored order of imports and re-arranged getters and setters

damianszczepanik commented:
Looks like we have now the most wanted feature - thank you!

kapoorlakshya commented:
Thank you, @KamikaZeeFu and @damianszczepanik! This is awesome!

damianszczepanik commented:
@KamikaZeeFu can you please update README file as more developers ask how to configure and use it?

KamikaZeeFu commented:
See #166 , hope that covers a few more use cases.

sumitg1 commented:
I am using the following script

node {
stage 'Checkout'
git url: 'https://github.com/sumitg1/selenium-cucumber-framework.git'

def mvnHome = tool 'MAVEN_HOME'

stage 'Test'

bat "${mvnHome}/bin/mvn test -Dcucumber.options=\"--tags @search --plugin json:target/first.json\" -Dwebdriver.driver=firefox"
  
stage "Publish First Run Report"
echo "***** Publish Reports *****"
step([$class: 'CucumberReportPublisher', jsonReportDirectory:'./target/', jenkinsBasePath: '',fileIncludePattern:'first.json', parallelTesting:true])


stage 'Retest'
echo "***** Retesting Failed TestCases  *****"
bat "${mvnHome}/bin/mvn test  -Dcucumber.options=\"@target\\rerun.txt --plugin json:target/second.json\""

stage "Publish Second Run Report"
echo "***** Publish Reports *****"
step([$class: 'CucumberReportPublisher', jsonReportDirectory:'./target/', jenkinsBasePath: '',fileIncludePattern:'second.json', parallelTesting:true])
}

	The job get stuck and going in loop

![image](https://user-images.githubusercontent.com/19838311/29984309-7843e384-8f77-11e7-9985-0382b48dd480.png)

![image](https://user-images.githubusercontent.com/19838311/29984330-970f66ee-8f77-11e7-9017-eb8468dcdcd5.png)
